shared/src/proto/types.rs

```diff
@@ -51,54 +52,69 @@ pub struct SignedTxData {
     pub sig: common::Signature,
 }
 
-/// A generic signed data wrapper for Borsh encode-able data.
+/// A serialization method to provide to [`Signed`], such
+/// that we may sign serialized data.
+pub trait SignedSerialize<T> {
+    /// A byte vector containing the serialized data.
+    type Output: AsRef<[u8]>;
+
+    /// Encodes `data` as a byte vector,
+    /// with some arbitrary serialization method.
+    fn serialize(data: &T) -> Self::Output;
+}
+
+/// Tag type that indicates we should use [`BorshSerialize`]
+/// to sign data in a [`Signed`] wrapper.
+#[derive(Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
+pub struct SerializeWithBorsh;
+
+impl<T: BorshSerialize> SignedSerialize<T> for SerializeWithBorsh {
+    type Output = Vec<u8>;
+
+    fn serialize(data: &T) -> Vec<u8> {
+        data.try_to_vec()
+            .expect("Encoding data for signing shouldn't fail")
+    }
+}
+
+/// A generic signed data wrapper for serialize-able types.
+///
+/// The default serialization method is [`BorshSerialize`].
 #[derive(
     Clone, Debug, BorshSerialize, BorshDeserialize, Serialize, Deserialize,
 )]
-pub struct Signed<T: BorshSerialize + BorshDeserialize> {
+pub struct Signed<T, S = SerializeWithBorsh> {
```

Comment:
It'd be good if we could bound `S` here (if it's possible?) so that it's impossible to construct a `Signed` with an invalid tag, something like

```rust
pub struct Signed<T, S: SignedSerialize<T> = SerializeWithBorsh>
```

Reply:
we already tighten `Signed` at the impl block starting on line 148. you can only verify signatures of `Signed` instances with a tag which implements `SignedSerialize`. you can't use the struct itself to instantiate new values, since it has a private field, but you can use `Signed::new_from` instead. regardless, we hit the same wall, this instance can't do jack if `S` does not implement `SignedSerialize`.

tl;dr I don't think this is an issue